pkg_resources/__init__.py

```diff
@@ -74,6 +74,8 @@
 
 import _imp
 
+sys.path.append(os.path.dirname(__file__) + '/../setuptools/_vendor')
```

Comment:
We might end up with 2 duplicated entries in `sys.path` right? (one for importing `pkg_resources` and another one for importing `setuptools`).

I suppose that this is not too bad...

Reply:
And it could be even more if those modules end up getting reloaded. My feeling was that such situations are probably rare and the inefficiencies caused are acceptable, compared with the cost of making this one-liner more complicated.

Reply:
Isn't it a simple `if path not in sys.path` ?
```suggestion
_vendor_path = os.path.dirname(os.path.dirname(__file__)) + "/setuptools/_vendor"
if _vendor_path not in sys.path:
    sys.path.append(_vendor_path)
```

Reply:
Yes, it's "as simple" as that, but as you've demonstrated, that changes this from a one-line expression into a multi-line, conditional code block. And since this logic needs to be copy/pasted, I'm loathe to make it more than one line. If it is more than one line, I'm tempted to create a separate module to contain the shared logic, but I'm also disinclined to do that because it would require creating a new top-level module.

The reason to have it in one place or in a single expression is it resists being tweaked. It resists, for example, the temptation to add behavior around the different lines, whereas a single expression stands on its own (it's self-encapsulating).

We could keep it a single expression with something like:

```suggestion
sys.path.extend(((vendor_path := os.path.join(os.path.dirname(os.path.dirname(__file__)), "setuptools", "_vendor")) not in sys.path) * [vendor_path])  # fmt: skip
```

It's convoluted, strains readability, but at least remains on one line and one expression.

Allowing ruff to reformat it results in:

```suggestion
sys.path.extend(
    (
        (
            vendor_path := os.path.join(
                os.path.dirname(os.path.dirname(__file__)), "setuptools", "_vendor"
            )
        )
        not in sys.path
    )
    * [vendor_path]
)
```

I'm a little unsure about whether to use `os.path.join` or not. I agree, it's often preferable (best practice) to use `join`, but in this case, it's conceivable that the path could be inside a zipfile, in which cases only `posixpath.sep` is supported, and since Windows has supported forward slashes since at least DOS 2.1, I'm comfortable using them for something like this hack (that we hope to retire). I'm not explicitly supporting vendoring for zip files, but it would be nice if it happens to work.

Indeed, I did some testing and confirmed that by using forward-slashes (and replacing `..` with another call to `dirname`), the technique works to load vendored packages even in a zip file:

```
 setuptools debt/2825-devendor [1] @ git diff
diff --git a/pkg_resources/__init__.py b/pkg_resources/__init__.py
index b8cf28cdb..21e070406 100644
--- a/pkg_resources/__init__.py
+++ b/pkg_resources/__init__.py
@@ -74,7 +74,17 @@ from pkgutil import get_importer
 
 import _imp
 
-sys.path.append(os.path.dirname(__file__) + '/../setuptools/_vendor')
+sys.path.extend(
+    (
+        (
+            vendor_path := posixpath.join(
+                os.path.dirname(os.path.dirname(__file__)), 'setuptools', '_vendor'
+            )
+        )
+        not in sys.path
+    )
+    * [vendor_path]
+)
 
 # capture these to bypass sandboxing
 from os import utime
 setuptools debt/2825-devendor @ py -m zipfile -c pr.zip pkg_resources setuptools
 setuptools debt/2825-devendor @ env PYTHONPATH=pr.zip py -c "import sys; sys.path.remove(''); import pkg_resources" && echo done
<string>:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import sys; sys.path.remove(''); import pkg_resources
done
```

Replace `posixpath` with `ntpath` and the import fails. Interestingly, with `ntpath` the import only fails on Unix, so apparently the zip importer does support using backslashes inside the zipfile. Oh, yes, I reviewed the zipimporter logic recently and (to my surprise), it actually converts forward slashes to backslashes when reading zipfile entries on Windows.

Given those findings, I'm fine to use `os.path.join`, as much as I've come to dislike the pattern.

```suggestion
sys.path.extend(
    (
        (
            vendor_path := os.path.join(
                os.path.dirname(os.path.dirname(__file__)), 'setuptools', '_vendor'
            )
        )
        not in sys.path
    )
    * [vendor_path]
)
```

```suggestion
sys.path.extend(((vendor_path := os.path.join(os.path.dirname(os.path.dirname(__file__)), 'setuptools', '_vendor')) not in sys.path) * [vendor_path])  # fmt: skip
```

Let's go with that and use the same technique in both setuptools and pkg_resources.

Reply:
Done in 930ebe55b.

Reply:
The `\` in zipfile is something I wouldn't even have considered, thanks for testing. My concern about consistent path separator was just so you wouldn't end up with the same path twice, differing only by slashes (inconsistent path separators would still protect against module reloading)